File: Problem/By section/Tree/Problem 637/problem_637.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // Time: O(n), Space: O(n)
    public List<Double> averageOfLevels(TreeNode root) {
        if(root == null) return null;
        
        List<Double> ans = new LinkedList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while(!queue.isEmpty()){
            double sum = 0;
            // important, the size of a level is the size of the whole queue 
            // after poping all the stuff from the previous layer
            int this_level_nodes = queue.size();
            for(int i = 0; i < this_level_nodes; i++){
                TreeNode cur_node = queue.poll();
                sum += cur_node.val;
                if(cur_node.left != null){
                    queue.add(cur_node.left);
                }
                
                if(cur_node.right != null){
                    queue.add(cur_node.right);
                }
            }
            ans.add(sum/this_level_nodes);
        }
        
        return ans;
    }
}

File: Problem/By section/Tree/Problem 637/problem_637.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
**/
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // Time: O(n), Space: O(m) (maximual number of nodes in one level)
    vector<double> averageOfLevels(TreeNode* root) {
        if(!root) return {0.0};
        
        vector<double> avg;
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()) {
            int size = q.size();
            double sum = 0.0;
            for(int i = 0; i< size; i++) {
                auto temp = q.front();
                sum += (double)temp->val;
                q.pop();
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
            
            avg.push_back(((double)sum / size));
        }
        
        return avg;
        
    }
};

int main(int argc, char** argv) {
    Solution s;

    return 0;
}